### DISCUSSION TITLE
why is kanaries_api_key required in offline mode?

### DISCUSSION THREAD

**Discussion by wzherwa** [Kanaries/pygwalker#505] [General]
may I kindly ask on why the kanaries_api_key is required in privacy=offline mode? 
The offline mode is supposed to 

- neither send any data 
- nor make any api calls
so for which calls is the kanaries_api_key required? 

when kanaries_api_key not set and privacy=offline, I get the following error messages:
> ERROR: kanaries_api_key is not valid.
Please set kanaries_api_key first.
If you are not kanaries user, please register it from 'https://kanaries.net/home/access' 
Then refer 'https://github.com/Kanaries/pygwalker/wiki/How-to-get-api-key-of-kanaries%3F' to set kanaries_api_key. 

> ERROR:pygwalker.services.cloud_service:kanaries_api_key is not valid.
Please set kanaries_api_key first.
If you are not kanaries user, please register it from 'https://kanaries.net/home/access' 
Then refer 'https://github.com/Kanaries/pygwalker/wiki/How-to-get-api-key-of-kanaries%3F' to set kanaries_api_key.

**Comment by ObservedObserver** [comment]
It seems you are trying to use cloud feature in offline mode. 
1. Please be aware of "cloud" means the opposite of "offline", so if you want to use it offline, you cannot use any cloud feature, like Ask-to-Viz, Publish to cloud, Cloud computing.
2. In next release, we will make sure all those cloud features' buttons and input are disabled or removed when you set offline mode.

**Reply by wzherwa** [reply]
thank you for your prompt reply @ObservedObserver , I am fully aware of the distinction between cloud and offline - that's why I am asking, which functions/parameters, I've used that would invoke "cloud feature".

Here's my simple code (calling the "run_pygwalker" function and the "data" passed is a pandas dataframe)

```python
import streamlit as st

from pygwalker.api.streamlit import (
    StreamlitRenderer,
    init_streamlit_comm,
    get_streamlit_html,
)

st.set_page_config(layout="wide")

# Establish communication between pygwalker and Streamlit
init_streamlit_comm()


@st.cache_resource
def get_pyg_renderer(data) -> "StreamlitRenderer":
    return StreamlitRenderer(
        data,
        # env="Streamlit",
        # use_kernel_calc=True,
        # spec="./gw_config.json",
        # spec_io_mode="rw",
    )


def run_pygwalker(data):
    renderer = get_pyg_renderer(data)
    renderer.render_explore()
```

and here's the output from
```shell
pygwalker config --list

INFO: Running in offline mode. There might be newer releases available. Please check at https://github.com/Kanaries/pygwalker or https://pypi.org/project/pygwalker.
Current configuration:
{
    "privacy": "offline",
    "kanaries_token": ""
}
```

so my final question: which part invokes "cloud service" and how can I avoid that?

btw. the same setup runs on windows w/o that error, but on macOS I get this "missing api_key" error

**Reply by ObservedObserver** [reply]
Can you describe what buttons you clicked. There are some buttons / text input in the UI related cloud features. If you clicked those on UI, you might get the err.

If you didn't click any buttons and the err just popup the moment you start the streamlit app. Please provide detailed information about your dev env and pygwalker version.

**Reply by wzherwa** [reply]
I can't even click a button. as the pygwalker UI doesn't even show up in streamlit ...
- pygwalker 0.4.7
- streamlit 1.32.2
- pyautogen 0.2.9

trying to use the "run_pygwalker" function as autogen function. Autogen correctly calls the function, but the "renderer = get_pyg_renderer(data)" function returns the error code

**Reply by ObservedObserver** [reply]
Please provide your pygwalker version, code snippet which can reproduce the err (I tested the exact same code you provided, it works on both win and macos).

**Reply by wzherwa** [reply]
got it now ... the data passed was actually not a pandas data frame, but a string ... solved it via 
```python
def run_pygwalker(data):
    df = pd.read_csv(data, sep=",")
    renderer = get_pyg_renderer(df)
    renderer.render_explore()
```

but thank you very much @ObservedObserver , really responsive support - will recommend this lib!
